### DISCUSSION TITLE
[Support] TLS Error: tls-crypt unwrapping failed

### DISCUSSION THREAD

**Discussion by Natherul** [pivpn/pivpn#1412] [Support] [answered]
---
name: Support Request
about: Request support from PiVPN comunity
title: "TLS Error: tls-crypt unwrapping failed"
labels: ''
assignees: ''

---

## In raising this issue, I confirm the following:

`{please fill the checkboxes, e.g: [X]}`

- [x] I have read the [documentation](https://docs.pivpn.io)
- [x] I have read and understood the [contributors guide](https://github.com/pivpn/pivpn/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [] The issue I am reporting can be *is* directly related to the pivpn installer script.

### Have you searched for similar issues and solutions?
    yes, searching for the error in title

### Has your install failed? 
```
no
```

#### What type of hardware are you running PiVPN at?

```
Raspberrypi 3 model B

```
#### Output of `uname -a`
```
Linux hostname 5.10.52-v7+ #1441 SMP Tue Aug 3 18:10:09 BST 2021 armv7l GNU/Linux
```

#### Output of `cat /etc/os-release`
```
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

### Issue
OpenVPN Connect on my phone does not work. I had an installation working but then I have not used it for more than a year and since then I have updated my RPI to bullseye and now it will not connect

### Console output of      `pivpn debug`

```
pi@lab ~ $ pivpn -d
::: Generating Debug Output
::::            PiVPN debug              ::::
=============================================
::::            Latest commit            ::::
Branch: test
Commit: 65ffa9afd262d68afd242823ccb0711c1a280fc8
Author: Orazio
Date: Thu Nov 18 14:28:35 2021 +0100
Summary: Spelling
=============================================
::::        Installation settings        ::::
PLAT=Raspbian
OSCN=bullseye
USING_UFW=0
IPv4dev=wlan0
dhcpReserv=1
IPv4addr=192.168.0.3/24
IPv4gw=192.168.0.1
install_user=pi
install_home=/home/pi
VPN=openvpn
pivpnPROTO=udp
pivpnPORT=1194
pivpnDNS1=8.8.8.8
pivpnDNS2=8.8.4.4
pivpnSEARCHDOMAIN=
pivpnHOST=REDACTED
TWO_POINT_FOUR=1
pivpnENCRYPT=256
USE_PREDEFINED_DH_PARAM=
INPUT_CHAIN_EDITED=0
FORWARD_CHAIN_EDITED=0
pivpnDEV=tun0
pivpnNET=10.8.0.0
subnetClass=24
ALLOWED_IPS=""
UNATTUPG=0
INSTALLED_PACKAGES=(openvpn)
HELP_SHOWN=1
=============================================
::::  Server configuration shown below   ::::
dev tun
proto udp
port 1194
ca /etc/openvpn/easy-rsa/pki/ca.crt
cert /etc/openvpn/easy-rsa/pki/issued/lab_d334b775-ed99-492a-8ad4-f94ace60641b.crt
key /etc/openvpn/easy-rsa/pki/private/lab_d334b775-ed99-492a-8ad4-f94ace60641b.key
dh none
ecdh-curve prime256v1
topology subnet
server 10.8.0.0 255.255.255.0
# Set your primary domain name server address for clients
push "dhcp-option DNS 8.8.8.8"
push "dhcp-option DNS 8.8.4.4"
# Prevent DNS leaks on Windows
push "block-outside-dns"
# Override the Client default gateway by using 0.0.0.0/1 and
# 128.0.0.0/1 rather than 0.0.0.0/0. This has the benefit of
# overriding but not wiping out the original default gateway.
push "redirect-gateway def1"
client-to-client
client-config-dir /etc/openvpn/ccd
keepalive 15 120
remote-cert-tls client
tls-version-min 1.2
tls-crypt /etc/openvpn/easy-rsa/pki/ta.key
cipher AES-256-CBC
auth SHA256
user openvpn
group openvpn
persist-key
persist-tun
crl-verify /etc/openvpn/crl.pem
status /var/log/openvpn-status.log 20
status-version 3
syslog
verb 3
#DuplicateCNs allow access control on a less-granular, per user basis.
#Remove # if you will manage access by user instead of device. 
#duplicate-cn
# Generated for use by PiVPN.io
=============================================
::::  Client template file shown below   ::::
client
dev tun
proto udp
remote REDACTED 1194
resolv-retry infinite
nobind
remote-cert-tls server
tls-version-min 1.2
verify-x509-name lab_d334b775-ed99-492a-8ad4-f94ace60641b name
cipher AES-256-CBC
auth SHA256
auth-nocache
verb 3
=============================================
::::    Recursive list of files in       ::::
::: /etc/openvpn/easy-rsa/pki shows below :::
/etc/openvpn/easy-rsa/pki/:
ca.crt
crl.pem
Default.txt
ecparams
index.txt
index.txt.attr
index.txt.attr.old
index.txt.old
issued
openssl-easyrsa.cnf
phone.ovpn
private
renewed
revoked
safessl-easyrsa.cnf
serial
serial.old
ta.key

/etc/openvpn/easy-rsa/pki/ecparams:
prime256v1.pem

/etc/openvpn/easy-rsa/pki/issued:
lab_d334b775-ed99-492a-8ad4-f94ace60641b.crt
phone.crt

/etc/openvpn/easy-rsa/pki/private:
ca.key
lab_d334b775-ed99-492a-8ad4-f94ace60641b.key
phone.key

/etc/openvpn/easy-rsa/pki/renewed:
private_by_serial
reqs_by_serial

/etc/openvpn/easy-rsa/pki/renewed/private_by_serial:

/etc/openvpn/easy-rsa/pki/renewed/reqs_by_serial:

/etc/openvpn/easy-rsa/pki/revoked:
private_by_serial
reqs_by_serial

/etc/openvpn/easy-rsa/pki/revoked/private_by_serial:
79E40C6979C408894784489F3A431366.key

/etc/openvpn/easy-rsa/pki/revoked/reqs_by_serial:
79E40C6979C408894784489F3A431366.req
=============================================
::::            Self check               ::::
:: [OK] IP forwarding is enabled
:: [OK] Iptables MASQUERADE rule set
:: [OK] OpenVPN is running
:: [OK] OpenVPN is enabled (it will automatically start on reboot)
:: [OK] OpenVPN is listening on port 1194/udp
=============================================
:::: Having trouble connecting? Take a look at the FAQ:
:::: https://docs.pivpn.io/faq
=============================================
::::      Snippet of the server log      ::::
Nov 21 13:23:56 lab ovpn-server[573]: REDACTED:55437 TLS Error: TLS handshake failed
Nov 21 13:23:56 lab ovpn-server[573]: REDACTED:55437 SIGUSR1[soft,tls-error] received, client-instance restarting
Nov 21 14:48:33 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:33 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:34 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:34 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:35 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:35 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:36 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:36 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:37 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:37 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:38 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:38 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:39 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:39 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:40 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:40 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
Nov 21 14:48:41 lab ovpn-server[573]: tls-crypt unwrap error: packet authentication failed
Nov 21 14:48:41 lab ovpn-server[573]: TLS Error: tls-crypt unwrapping failed from [AF_INET]REDACTED:58387
=============================================
::::            Debug complete           ::::
::: 
::: Debug output completed above.
::: Copy saved to /tmp/debug.log
::: 

```
### Have you taken any steps towards solving your issue?
```
Tried to reinstall several times, both on normal and testing branch
```

**Comment by Natherul** [comment]
I did not check the box if it sdirectly related to the script because Im not sure. Since a reinstall does the same I think that it is but I cannot answer with certainty.

**Comment by coolapso** [comment]
Moving to discussions, this is connection related. please check other issues and discussions, you will find plenty of those your solution might be in one of those, if not then feel free to reply back. 

Moving to discussions as per: 

https://github.com/pivpn/pivpn/blob/master/README.md#general-guidelines

> The Github issues page is not intended for general support usage or feature requests
PiVPN team may close any Discussion/Issue without any warning if they don't follow the Guidelines

**Comment by Natherul** [comment]
I missed discussions entirely however, there is no topic in discussions that fits either. When I searched first I did find:
https://github.com/pivpn/pivpn/issues/639
but its closed with no clear answer on how to fix hence me starting the new report. I have so far gotten nowhere.

**Reply by coolapso** [reply]
No worries, if there's no new discussion you can just open one :) 

heres a list of all issues contianing "unwrap errors" https://github.com/pivpn/pivpn/issues?q=unwrap+error

**Reply by coolapso** [reply]
In a nutshell that error is very often related with connectivity / client sided, 

Where are you trying to connect from? 
what client? 
What device and OS?

**Reply by coolapso** [reply]
Sorry I ntoiced you filled that up there.. but still need more data and you are using the wrong template, you are assuming its a bug when it is not. 

Please edit the original topic,  use the following template and add the requested data: https://raw.githubusercontent.com/pivpn/pivpn/master/.github/ISSUE_TEMPLATE/support-request.md

**Reply by Natherul** [reply]
> Sorry I ntoiced you filled that up there.. but still need more data and you are using the wrong template, you are assuming its a bug when it is not.
> 
> Please edit the original topic, use the following template and add the requested data: https://raw.githubusercontent.com/pivpn/pivpn/master/.github/ISSUE_TEMPLATE/support-request.md

Updated. I assumed it was a bug initially since a clean install still comes up with the issue.

**Comment by Natherul** [comment, marked as answer]
Reinstalled again and used TCP instead of UDP and then it all works.... So there is something rotten with UDP settings.

**Reply by coolapso** [reply]
Maybe you didn't forwarded the correct protocol on your router? that could be one of the reasons why.

**Reply by Natherul** [reply]
> Maybe you didn't forwarded the correct protocol on your router? that could be one of the reasons why.

Nope. Thats def port routed right. And I should not have gotten to the logs then?
Either way im 100% sure the port forwarding is/was right

**Comment by zuhairmahd** [comment]
Experiencing this issue with UDP.  If I take out the TLS key from the server by commenting 
tls-crypt /etc/openvpn/easy-rsa/pki/ta.key on the server and remove the <tls-crypt> ... </tls-crypt> from the client all works fine.
Worth mentioning that:
1. This does not happen with everyone.  I have family connecting to my server from other countries and all works fine.
2. This, so far, only happens when I try to connect using the iOS client.  No issues with Android or PC.
3. This issue started as being intermittent a few weeks ago, but stopped working all together last week.
4. This is actually happening on two seperate machines, one running in AWS and the other is a PI running at home -- totally seperate networks and everything.
Here is what appears on the server when I try to connect with TLS Key Negotiation enabled:
Aug 21 19:27:13 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29328 tls-crypt unwrap error: bad packet ID (may be a replay): [ #1 / time = (1661110018) Sun Aug 21 19:26:58 2022 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Aug 21 19:27:13 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29328 tls-crypt unwrap error: packet replay
Aug 21 19:27:13 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29328 TLS Error: tls-crypt unwrapping failed from [AF_INET]86.108.20.251:29328
Aug 21 19:27:14 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29308 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Aug 21 19:27:14 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29308 TLS Error: TLS handshake failed
Aug 21 19:27:14 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29308 SIGUSR1[soft,tls-error] received, client-instance restarting
Aug 21 19:27:43 ip-172-26-10-251 vmunix: [651770.662971] device eth0 left promiscuous mode
Aug 21 19:27:58 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29328 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Aug 21 19:27:58 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29328 TLS Error: TLS handshake failed
Aug 21 19:27:58 ip-172-26-10-251 ovpn-server[223583]: 86.108.20.251:29328 SIGUSR1[soft,tls-error] received, client-instance restarting

After a few rounds of this, the iOS client gives up and disconnects.

Very strange.

**Comment by TinCanTech** [comment]
If TCP works and UDP does **not** then the network you are using is unreliable.


UDP is a connectionless protocol, with no guarantees that packets will arrive at their destination. They can arrive _out of order_ or even _not at all_. That is the nature of UDP.

TCP is connection-oriented protocol, which provides reliable, ordered, and error-checked delivery.

**Comment by zuhairmahd** [comment]
@TinCanTech  UDP works beautifully here, just not with TLS negotiation. :-)

**Reply by TinCanTech** [reply]
@zuhairmahd Your log shows that UDP packets are not being delivered reliably: 
- `tls-crypt unwrap error: packet replay`

TCP is one solution. Disabling the TLS key is not such a good idea ..

**Reply by zuhairmahd** [reply]
What's fascinating though is this is only happening on iOS clients -- android and windows work just fine, which to me rules out an ISP issue (though not entirely). I rather suspect the iOS client, since the issue occurs whether it is connected to the Wi-Fi network or 4G (though both are powered by Orange). The only other significant event that took place in the past couple of weeks was an update to the iOS client. Since I'm the only one having problems, I'll just spin off a separate VM and configure it to work for me, instead of changing to TCPIP and having to reconfigure everyone else.  Oh and yes, I did reinstall the client and did do a "reset Network Settings" on my iPhone SE 2022.

**Reply by Pi-DD** [reply]
I have exactly this problem - on my iPhone (6s, iOS 15.7.1) but not on my iPad (8th Gen, iOS 16.1.1) which works just fine from exactly the same spot on exactly the same wifi network. The "settings" for the two clients are identical
The latest Win 10 client will not connect either. 

Any sign of some sort of fix?

**Reply by TinCanTech** [reply]
@Pi-DD `Me2` posts don't help. Details are required.
